pkg/hub/http_user_server.go

```diff
@@ -0,0 +1,194 @@
+	"sigs.k8s.io/apiserver-network-proxy/pkg/util"
+)
+
+type HTTPUserServer struct {
+	getTunnel       func() (konnectivity.Tunnel, error)
+	proxyServerHost string
+	proxyServerPort int
+
+	proxyCACertPath, proxyCertPath, proxyKeyPath string
+
+	serverCert, serverKey string
+	serverPort            int
+}
+
+func (k *HTTPUserServer) AddFlags(cmd *cobra.Command) {
+	flags := cmd.Flags()
+
+	flags.StringVar(&k.proxyServerHost, "host", k.proxyServerHost, "The host of the ANP proxy-server")
+	flags.IntVar(&k.proxyServerPort, "port", k.proxyServerPort, "The port of the ANP proxy-server")
+
+	flags.StringVar(&k.proxyCACertPath, "proxy-ca-cert", k.proxyCACertPath, "The path to the CA certificate of the ANP proxy-server")
+	flags.StringVar(&k.proxyCertPath, "proxy-cert", k.proxyCertPath, "The path to the certificate of the ANP proxy-server")
+	flags.StringVar(&k.proxyKeyPath, "proxy-key", k.proxyKeyPath, "The path to the key of the ANP proxy-server")
+
+	flags.StringVar(&k.serverCert, "server-cert", k.serverCert, "Secure communication with this cert")
+	flags.StringVar(&k.serverKey, "server-key", k.serverKey, "Secure communication with this key")
+	flags.IntVar(&k.serverPort, "server-port", k.serverPort, "handle user request using this port")
+}
+
+func (k *HTTPUserServer) Validate() error {
+	if k.serverCert == "" {
+		return fmt.Errorf("The server-cert is required")
+	}
+
+	if k.serverKey == "" {
+		return fmt.Errorf("The server-key is required")
+	}
+
+	if k.serverPort == 0 {
+		return fmt.Errorf("The server-port is required")
+	}
+
+	return nil
+}
+
+func NewHTTPUserServer() *HTTPUserServer {
+	return &HTTPUserServer{}
+}
+
+func (k *HTTPUserServer) init(ctx context.Context) error {
+	proxyTLSCfg, err := util.GetClientTLSConfig(k.proxyCACertPath, k.proxyCertPath, k.proxyKeyPath, k.proxyServerHost, nil)
+	if err != nil {
+		return err
+	}
+	k.getTunnel = func() (konnectivity.Tunnel, error) {
+		// instantiate a gprc proxy dialer
+		tunnel, err := konnectivity.CreateSingleUseGrpcTunnel(
+			ctx,
+			net.JoinHostPort(k.proxyServerHost, strconv.Itoa(k.proxyServerPort)),
+			grpc.WithTransportCredentials(grpccredentials.NewTLS(proxyTLSCfg)),
+		)
+		if err != nil {
+			return nil, err
+		}
+		return tunnel, nil
+	}
+	return nil
+}
+
+func (k *HTTPUserServer) handler(wr http.ResponseWriter, req *http.Request) {
+	if klog.V(4).Enabled() {
+		dump, err := httputil.DumpRequest(req, true)
+		if err != nil {
+			http.Error(wr, err.Error(), http.StatusBadRequest)
+			return
+		}
+		klog.V(4).Infof("request:\n%s", string(dump))
+	}
+
+	// parse clusterID from current requestURL
+	clusterID, kubeAPIPath, err := parseRequestURL(req.RequestURI)
+	if err != nil {
+		http.Error(wr, err.Error(), http.StatusBadRequest)
+		return
+	}
+
+	target := fmt.Sprintf("https://%s", clusterID)
+	apiserverURL, err := url.Parse(target)
+	if err != nil {
+		http.Error(wr, err.Error(), http.StatusBadRequest)
+		return
+	}
+
+	tunnel, err := k.getTunnel()
```

Comment:
the tunnel seems to be same for each request, right? if it's right, why we create the tunnel every time?

Reply:
In our test, the tunnel is not supported reused many times. This should be an enhancement TODO in the next step.

Reply:
Add a comments here?

Reply:
Sure, the TODO comment added.